internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -507,6 +509,112 @@ func (a *Adapter) EnsureOverrideSnapshotValid() (controller.OperationResult, err
+		err := gitops.AnnotateSnapshot(a.context, a.snapshot, gitops.PRGroupCreationAnnotation, "Failed to get PR group from snapshot", a.client)
+		if err != nil {
+			return controller.RequeueWithError(err)
+		}
+		return controller.ContinueProcessing()
+	}
+
+	pipelineRuns, err := a.loader.GetPipelineRunsWithPRGroupHash(a.context, a.client, a.snapshot, prGroupHash)
+	if err != nil {
+		a.logger.Error(err, fmt.Sprintf("Failed to get build pipelineRuns for given pr group hash %s", prGroupHash))
+		return controller.RequeueWithError(err)
+	}
+
+	for _, pipelineRun := range *pipelineRuns {
+		pipelineRun := pipelineRun
+
+		// check if build PLR finishes
+		if !h.HasPipelineRunFinished(&pipelineRun) {
+			a.logger.Info(fmt.Sprintf("The build pipelineRun %s/%s with pr group %s is still running, won't create group snasphot", pipelineRun.Namespace, pipelineRun.Name, prGroup))
+			err := gitops.AnnotateSnapshot(a.context, a.snapshot, gitops.PRGroupCreationAnnotation, fmt.Sprintf("The build pipelineRun %s/%s with pr group %s is still running, won't create group snasphot", pipelineRun.Namespace, pipelineRun.Name, prGroup), a.client)
+			if err != nil {
+				return controller.RequeueWithError(err)
+			}
+			return controller.ContinueProcessing()
+		}
+
+		// check if build PLR succeeds
+		if !h.HasPipelineRunSucceeded(&pipelineRun) {
+			a.logger.Info(fmt.Sprintf("The build pipelineRun %s/%s with pr group %s failed, won't create group snasphot", pipelineRun.Namespace, pipelineRun.Name, prGroup))
+			err := gitops.AnnotateSnapshot(a.context, a.snapshot, gitops.PRGroupCreationAnnotation, fmt.Sprintf("The build pipelineRun %s/%s with pr group %s failed, won't create group snasphot", pipelineRun.Namespace, pipelineRun.Name, prGroup), a.client)
+			if err != nil {
+				return controller.RequeueWithError(err)
+			}
+			return controller.ContinueProcessing()
+		}
+
+		// check if build PLR has component snapshot created
+		if !metadata.HasAnnotation(&pipelineRun, tekton.SnapshotNameLabel) {
+			a.logger.Info(fmt.Sprintf("The build pipelineRun %s/%s with pr group %s has succeeded but component snapshot has not been created, won't create group snasphot ", pipelineRun.Namespace, pipelineRun.Name, prGroup))
+			err := gitops.AnnotateSnapshot(a.context, a.snapshot, gitops.PRGroupCreationAnnotation, fmt.Sprintf("The build pipelineRun %s/%s with pr group %s has succeeded but component snapshot has not been created, won't create group snasphot ", pipelineRun.Namespace, pipelineRun.Name, prGroup), a.client)
+			if err != nil {
+				return controller.RequeueWithError(err)
+			}
+			return controller.ContinueProcessing()
+		}
+	}
+
+	groupSnapshot, componentSnapshotInfos, err := a.prepareGroupSnapshot(a.application, prGroupHash)
+	if err != nil {
+		a.logger.Error(err, "Failed to prepare group snapshot")
+		return controller.RequeueWithError(err)
+	}
+
+	if groupSnapshot == nil || len(groupSnapshot.Spec.Components) < 2 {
```

Comment:
The group Snapshot should always contain all components of the application even if they weren't updated as part of the PR group - in that case they would just reference the GCL images.

Maybe this check would need to check the `GroupSnapshotInfoAnnotation` to check if there are less than two entries in it?

Reply:
Did I misunderstand your proposal to not have all components in [slack](https://redhat-internal.slack.com/archives/C034UFMUSV6/p1722867231577579?thread_ts=1722861891.339739&cid=C034UFMUSV6)?

Reply:
Read the discussion in [slack](https://redhat-internal.slack.com/archives/C034UFMUSV6/p1722867231577579?thread_ts=1722861891.339739&cid=C034UFMUSV6) again, I did misunderstand your meaning. That original `gclSnapshots` was to count the candidate of the pr component snapshots, and then add the left snasphotComponents from GCL, I thought you meant to not include the snapshotComponent from GCL. Let me add them back. Thanks!

Reply:
Yes, agreed.

Reply:
Updated. Thanks.